### PR TITLE
Support for self-registering slaves, refactoring

### PR DESCRIPTION
**Pull request by kshepitko**
- Support for self-registering slaves (the ones that register themselves in master - e.g., Swarm slaves). 
- Ability to explicitly mark the image as a slave image in pod template (old way of detecting if a JNLP slave is still supported)
- Unified approach for time-limited slave operations
- Refactoring performed
- Test coverage increased

### CODE REVIEW

**Comment by carlossg** [comment]
big PRs are hard to review, and mixing formatting/refactoring changes with new features is a problem.

How is a "self-registering slave" different from using the swarm plugin like in https://www.infoq.com/articles/scaling-docker-kubernetes-v1 ?
what does this PR add?

**Comment by kshepitko** [comment]
Hi Carlos,

Thanks for a quick reply!

I know the article you've referenced, I've read it a year ago. It's very good indeed! However, the approach you are talking about there is not as flexible as when using Swarm slave with Kubernetes plugin.
First of all, it's an aspect of scaling slaves up and down. Without plugin you either have to scale slave rc manually via kubectl, or rely on autoscaler which obviously doesn't work the same way as EMA algorithm in Jenkins node provisioning strategy, it's completely different. Basically autoscaler works based on the node load rather than on Jenkins load. 
Secondly, I have different slaves in Jenkins, so it's convenient to keep all slave configs in the same place.
Thirdly, I can define volume mounts and other perks for my Swarm slave pod right in the K8s plugin config, so I don't have to keep it in the K8s manifests outside of Jenkins.

The reason why I need exactly Swarm slave, not just plain JNLP, is that the slave image I use has some Java logic, and it's using Swarm client API to initiate slave process. Furthermore, in my situation it's impossible to have this Java logic in a Jenkins plugin so that theoretically a plain JNLP slave would be sufficient to carry out the job.

And here's the list of what this PR adds:

* Support for the self-registering slaves like Swarm. In this case the slave is not an instance of KubernetesSlave, it's just a Slave. Therefore the provisioning callback is a bit different: it creates the pod, waits until it comes online as a slave (as the pod is supposed to take care of its handshake with Jenkins), and then applies the retention strategy and updates the labels. So all parameters defined in pod/container template are primarily supported by this type of slave. User just has to ensure the pod name is equal to the slave name (it's easy to do with Swarm slave with `disableClientsUniqueId`).
* Possibility to define whether image represents slave or not with checkbox in UI (described in README). Current approach with "if the name is 'jnlp', it's a slave" doesn't seem to be the best - and won't work with Swarm slave images. However, this approach is still supported in the PR - so if you name your slave "jnlp", it will be assumed to be slave regardless whether you checked the checkbox or not. If you did, then "Self-registering agent" checkbox appears below. If checked, you can define a Swarm slave image and enjoy the perks of K8s plugin with it.

Other functional changes are just code improvements which I had to do in order to avoid copy/paste.

Apologies for the big PR! Basically the refactoring was done primarily in KubernetesCloud, which is a bit of a god class. I tried to make as less changes as possible there. The reason why that was necessary is that I have added a separate provisioning callback for self-registering slaves, and I had to extract common logic from the basic callback into smaller methods to avoid copy/paste. What I'd like to do as well is to extract some stuff from KubernetesCloud to outer classes (e.g., provisioning callbacks) to decrease its size.

Some changes are caused by different order of imports in my IDE. I think that code style XML for IDE would be a beneficial thing to have, to avoid this problem. Haven't found it in your project, but if you can point me to such anywhere, I'd be happy to update the pull request, to reduce the number of changes.

Please let me know if you have more questions.

Regards,
Kirill.

**Comment by carlossg** [comment]
> that the slave image I use has some Java logic

why not extend the docker image for the jnlp slave to do that logic, or use the command parameter to launch a different container when the pod is created?

**Comment by kshepitko** [comment]
I think it really depends on the architecture you decide to choose.

In our case there's a variety of different Swarm slaves doing specific things for the master, and are functionally bound to a proprietary Jenkins plugin by a specialized SPI "worker" interface; plugin just delegates work to the slaves, and each label serves a particular purpose. The bottom line is that in this case the logic has to be present in the same JVM with the slave (otherwise SPI won't work). So the delegation to a different container wouldn't work since SPI wouldn't work. 

Maybe this is a specific case, but having this proprietary plugin (which we can't really change) working together with the patched K8s plugin that supports Swarm slaves fit out needs perfectly. 

In case of such K8s plugin people can choose to put their logic in a different container, or use Swarm client API and have it all in one - so there's a flexibility here. In case if only plain JNLP slaves are supported, plugin is actually limiting you to particular architectural decisions (which is sometimes impossible, like in the case described above).

**Comment by carlossg** [comment]
I'm not talking about having a pod with two containers and delegating to another, I'm talking about extending the `jenkinsci/jnlp-slave` docker image and adding your custom logic in there. Then the plugin would start a pod from your image and execute a command in there (that can also be customized)
Why wouldn't that work for you?

**Comment by kshepitko** [comment]
Do you mean extracting the logic into a separate JAR and adding in it to `java -cp` of the entrypoint script (https://github.com/jenkinsci/docker-jnlp-slave/blob/master/jenkins-slave)? 
That would work actually. However, the refactoring effort in our case would be significant.

**Comment by carlossg** [comment]
Let's see if I'm getting it right:

* you have a docker image that on start connects to jenkins
* so you must have some command that already does the connection from the container to jenkins

why you don't 

* set the jnlp image in podTemplate to the image you already have
* execute the command you have to connect either in ENTRYPOINT or using podTemplate container `cmd` argument

**Comment by kshepitko** [comment]
> you have a docker image that on start connects to jenkins

Correct. Container accepts some parameters like master location, etc. - and Java app inside it passes the params to the Swarm client API, which talks to the master (where Swarm plugin is installed), and Swarm plugin creates an agent in Jenkins.

> why you don't

That's what I actually started with. Defined our Swarm slave image in Pod template and set the init params for successful agent start.
However, as a result there were 2 agents trying to connect to Jenkins: one represented by KubernetesSlave instance connected to JNLP agent container that the K8s plugin adds to the pod, and another being a Swarm agent (useless, since JNLP agent was taking the job that triggered the agent rollout, as the job is tied to the label) connected to Swarm container. If I set the Swarm options so that Swarm agent has the same name as the JNLP agent, there's only one of them remaining, a winner in competition:

`Aug 18, 2017 11:26:41 AM INFO org.csanchez.jenkins.plugins.kubernetes.KubernetesCloud$ProvisioningCallback call
Waiting for Pod to be scheduled (0/100): swarm-slave-template-mff79
Aug 18, 2017 11:26:47 AM INFO hudson.TcpSlaveAgentListener$ConnectionHandler run
Accepted connection #5 from 172.30.94.5/172.30.94.5:57597
Aug 18, 2017 11:26:47 AM INFO hudson.TcpSlaveAgentListener$ConnectionHandler run
Accepted connection #6 from 172.30.94.5/172.30.94.5:57598
Aug 18, 2017 11:26:48 AM SEVERE org.csanchez.jenkins.plugins.kubernetes.KubernetesCloud$ProvisioningCallback call
Error in provisioning; slave=KubernetesSlave name: swarm-slave-template-mff79, template=org.csanchez.jenkins.plugins.kubernetes.PodTemplate@7dac281: Node was deleted, computer is null
Aug 18, 2017 11:26:50 AM WARNING jenkins.slaves.DefaultJnlpSlaveReceiver channelClosed
Computer.threadPoolForRemoting [#38] for swarm-slave-template-mff79 terminated
Aug 18, 2017 11:26:50 AM WARNING hudson.util.ExceptionCatchingThreadFactory uncaughtException
Thread Computer.threadPoolForRemoting [#38] terminated unexpectedly
java.lang.IllegalStateException: Already connected
	at hudson.slaves.SlaveComputer.setChannel(SlaveComputer.java:567)
	at jenkins.slaves.DefaultJnlpSlaveReceiver.afterChannel(DefaultJnlpSlaveReceiver.java:172)
......
Aug 18, 2017 11:26:50 AM WARNING hudson.slaves.NodeProvisioner$2 run
Unexpected exception encountered while provisioning agent Kubernetes Pod Template
java.lang.IllegalStateException: Node was deleted, computer is null
	at org.csanchez.jenkins.plugins.kubernetes.KubernetesCloud$ProvisioningCallback.call(KubernetesCloud.java:793)
......`

I tried it a few times, and even considering that JNLP agent "lost battle" every time, this is not a reliable way to go.

And the scaling down doesn't work: 
- Swarm agent isn't auto-deleted after time defined in "Time in minutes to retain slave when idle", since retention strategy was defined for KubernetesSlave, not for it;
- Swarm agent can be deleted from Jenkins manually by clicking "Delete agent", but it doesn't delete the pod from the K8s.

All of these problems are not actual in the current PR.

**Comment by carlossg** [comment]
if you name your swarm container `jnlp` then it won't create two agents

**Comment by kshepitko** [comment]
1) It will still try to create 2 nodes in Jenkins anyways, even having one container (one is KubernetesSlave and another one created by Swarm plugin).
2) It's a hack, and it looks nasty:

```
Waiting for Pod to be scheduled (0/100): swarm-slave-template-fq6nk
Aug 21, 2017 10:33:46 AM INFO hudson.TcpSlaveAgentListener$ConnectionHandler run
Accepted connection #26 from 172.30.94.5/172.30.94.5:47282
Aug 21, 2017 10:33:48 AM SEVERE org.csanchez.jenkins.plugins.kubernetes.KubernetesCloud$ProvisioningCallback call
Error in provisioning; slave=KubernetesSlave name: swarm-slave-template-fq6nk, template=org.csanchez.jenkins.plugins.kubernetes.PodTemplate@774d6726: Node was deleted, computer is null
Aug 21, 2017 10:33:50 AM WARNING hudson.slaves.NodeProvisioner$2 run
Unexpected exception encountered while provisioning agent Kubernetes Pod Template
java.lang.IllegalStateException: Node was deleted, computer is null
	at org.csanchez.jenkins.plugins.kubernetes.KubernetesCloud$ProvisioningCallback.call(KubernetesCloud.java:793)
........
```

So each time when a new slave needs to be rolled out a KubernetesSlave instance is removed from Jenkins by K8s plugin when it fails to find the attached Computer. 
And automatic scale down still won't work.

File: README.md [hunk 1]
```diff
@@ -263,11 +264,23 @@ Example at [examples/declarative.groovy](examples/declarative.groovy)
...
+Alternatively, if you have a custom agent image, just check `A Jenkins agent image` check box. 
 
 Other containers must run a long running process, so the container does not exit. If the default entrypoint or command
 just runs something and exit then it should be overriden with something like `cat` with `ttyEnabled: true`.
 
+Please note that you are allowed to have only one agent container per pod template.
+
+# Self-registering agents
+
+If your slave container is capable of self-registering itself in Jenkins as a slave when the pod starts (for instance, if you use a
```
Comment (marvinthepa) [inline README.md]: Either `self-registering`, or `registering itself` - not both.

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/ContainerTemplateTestUtils.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    }
+
+    public static ContainerTemplate containerTemplate(String name, boolean slaveImage) {
+        return containerTemplate(name, slaveImage, false);
+    }
+
+    public static ContainerTemplate containerTemplate(String name, boolean slaveImage, boolean selfRegisteringSlave) {
+        ContainerTemplate result = new ContainerTemplate(name, "image");
+        result.setSlaveImage(slaveImage);
+        result.setSelfRegisteringSlave(slaveImage & selfRegisteringSlave);
```
Comment (marvinthepa) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/ContainerTemplateTestUtils.java]: Why are you using bit-wise and (`&`) on a boolean? Did you mean `&&`?

**Comment by carlossg** [comment]
maybe I didn't explain myself clearly, I'm not saying that you should use swarm, because on the jenkins side the swarm plugin will create a node when receiving the connection IIRC.
I'm saying that whatever code you use to configure the swarm agent, you can use it to configure the agent started by this plugin because you have full access to customize what command is called in the agent container.

**Comment by marvinthepa** [comment]
@carlossg: I think the problem is that the jnlp agent is very thin - as far as I understand it it just downloads the "real" agent from the Jenkins master and then executes it.

The code that @kshepitko uses needs to be in the same JVM (see https://github.com/jenkinsci/kubernetes-plugin/pull/200#issuecomment-322998349), so it would have to be IN the JNLP slave, which is downloaded from the master.

According to https://github.com/jenkinsci/kubernetes-plugin/pull/200#issuecomment-322781043, a Jenkins plugin won't do:

> The reason why I need exactly Swarm slave, not just plain JNLP, is that the slave image I use has some Java logic, and it's using Swarm client API to initiate slave process. Furthermore, in my situation it's impossible to have this Java logic in a Jenkins plugin so that theoretically a plain JNLP slave would be sufficient to carry out the job.

I.e. he has a complete custom slave jar, not just needs to execute a few commands before initiating the JNLP process.

correct, @kshepitko ?

I guess another solution would be to somehow make Jenkins provide a custom slave jar via JNLP, but I don't know if that is feasible...

File: README.md (hunk 1 above)
Comment (kshepitko) [inline README.md]: Oops, my bad!

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/ContainerTemplateTestUtils.java (hunk 2 above)
Comment (kshepitko) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/ContainerTemplateTestUtils.java]: True, it was bitwise on purpose before refactoring - now it should be `&&` indeed, right you are.

**Comment by kshepitko** [comment]
> I'm saying that whatever code you use to configure the swarm agent, you can use it to configure the agent started by this plugin because you have full access to customize what command is called in the agent container.

Sure @carlossg , and that's what I'm actually doing to make it connecting to Jenkins master - passing arguments like master address, disable unique hash, etc. But that cannot prevent the creation and attachment of obsolete KubernetesSlave to master, and that doesn't make Swarm agent's lifecycle controllable by K8s plugin (scaling down doesn't happen).
Does this make sense?

**Comment by kshepitko** [comment]
> he has a complete custom slave jar, not just needs to execute a few commands before initiating the JNLP process.

That's correct, yes. Carlos's suggestion to put this JAR on the classpath of the JNLP slave makes sense, but just doesn't apply in my case due to several reasons, one of those being major code refactoring needed. One of the things that would be refactored is different maintenance tasks that get initiated in the Swarm agent now on JVM startup. To make it work with JAR-on-JNLP-agent-classpath approach I need to wrap Main class into my own adapter to call it. Also, I need to rewrite the JNLP slave startup script in any case. And that's only a few reasons. So it's really easier to use K8s plugin with Swarm agent support in my case.

However, I'll try the suggested JAR-on-JNLP-agent-classpath approach very soon, to be used in new use cases.

So the pull request is about supporting Swarm slaves, when for some reasons it's impossible to use plain JNLP slaves.

**Comment by Vlatombe** [comment]
This is stale. Feel free to reopen after updating this PR.

**Pull request closed without merging**